File: benchmarks/public_suite/cpu_matrix/matrix_naive.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func matrixMultNaive(n int) float64 {
	a := make([][]float64, n)
	b := make([][]float64, n)
	c := make([][]float64, n)
	for i := 0; i < n; i++ {
		a[i] = make([]float64, n)
		b[i] = make([]float64, n)
		c[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			a[i][j] = 1.0
			b[i][j] = 2.0
		}
	}

	for i := 0; i < n; i++ {
		for k := 0; k < n; k++ {
			aik := a[i][k]
			for j := 0; j < n; j++ {
				c[i][j] += aik * b[k][j]
			}
		}
	}
	return c[0][0] + c[n-1][n-1]
}

func main() {
	n := 500
	if len(os.Args) > 1 {
		if val, err := strconv.Atoi(os.Args[1]); err == nil {
			n = val
		}
	}

	t1 := time.Now()
	checksum := matrixMultNaive(n)
	ms := float64(time.Since(t1).Nanoseconds()) / 1e6

	fmt.Printf("{\"lang\": \"Go (Naive)\", \"test\": \"Matrix 500x500\", \"checksum\": %.1f, \"time_ms\": %.3f}\n", checksum, ms)
}

File: benchmarks/public_suite/cpu_matrix/matrix_naive.rs
use std::env;
use std::time::Instant;

fn matrix_mult_naive(n: usize) -> f64 {
    let a = vec![vec![1.0f64; n]; n];
    let b = vec![vec![2.0f64; n]; n];
    let mut c = vec![vec![0.0f64; n]; n];

    for i in 0..n {
        for k in 0..n {
            let aik = a[i][k];
            for j in 0..n {
                c[i][j] += aik * b[k][j];
            }
        }
    }
    c[0][0] + c[n - 1][n - 1]
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = if args.len() > 1 { args[1].parse().unwrap_or(500) } else { 500 };

    let t1 = Instant::now();
    let checksum = matrix_mult_naive(n);
    let ms = t1.elapsed().as_secs_f64() * 1000.0;

    println!("{{\"lang\": \"Rust (Naive)\", \"test\": \"Matrix 500x500\", \"checksum\": {}, \"time_ms\": {:.3}}}", checksum, ms);
}
